### PR TITLE
feat: add `StreamingChunk` to Cohere ChatGenerator 

### LINKED ISSUE #2071: feat: Update `CohereChatGenerator` to use the `StreamingChunk` fields
Refer to parent issue for more details

### PR DESCRIPTION
**Pull request by davidsbatista**
### Related Issues

- fixes #2071 

### Proposed Changes:

- Replaced current streaming parsing with a `_convert_cohere_chunk_to_streaming_chunk()` that converts Cohere chunks to Haystack's StreamingChunk format.
- Using the `_convert_streaming_chunks_to_chat_message` from `haystack.components.generators.utils`
- Added imports for FinishReason and ToolCallDelta from haystack.dataclasses.streaming_chunk.
- Updated `haystack-ai` dependency from `>=2.15.1` to `>=2.16.1` to leverage new streaming utilities.

### How did you test it?

- Added tests for `_convert_cohere_chunk_to_streaming_chunk()`  covering all chunk types and edge cases.

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 1]
```diff
@@ -359,15 +393,75 @@ async def _parse_async_streaming_response(
...
+            chunk=chunk, previous_chunks=chunks, component_info=component_info, model=model
+        )
+        chunks.append(streaming_chunk)
+        await streaming_callback(streaming_chunk)
+
+    return _convert_streaming_chunks_to_chat_message(chunks=chunks)
+
+
+def _convert_streaming_chunks_to_chat_message(chunks: List[StreamingChunk]) -> ChatMessage:
+    """
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: Why are you implementing this function here? and not using [ prebuilt function](https://github.com/deepset-ai/haystack/blob/5f3c37d287a4fa882d59161a20f762e9102bc6aa/haystack/components/generators/utils.py#L71). Does it need some additional logic?

Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: I was not even aware of that function - I will look into it, probably can be used as is.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 2]
```diff
@@ -190,144 +176,189 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-    if not hasattr(cohere_chunk, "delta") or cohere_chunk.delta is None:
-        return None
+    if not chunk or not hasattr(chunk, "delta") or chunk.delta is None:
+        return StreamingChunk(
+            content="",
+            component_info=component_info,
+            index=None,
+            finish_reason=None,
+            meta={"model": model},
+        )
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: I would suggest saving the intended param values in if conditions and constructing a `StreamingChunk` once at the end for code simplification.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 3]
```diff
@@ -190,144 +177,139 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-    :return: Content to stream (if any), None otherwise
+    :returns:
+        A StreamingChunk object representing the content of the chunk from the Cohere API.
     """
-    if not cohere_chunk:
-        return None
+    finish_reason_mapping: Dict[str, FinishReason] = {
+        "COMPLETE": "stop",
+        "MAX_TOKENS": "length",
+        "ERROR": "content_filter",
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: I think we can skip mapping Error since we don’t have a suitable mapping value. In that case, it will simply return None.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 4]
```diff
@@ -190,144 +177,139 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-            return content
-    elif cohere_chunk.type == "tool-plan-delta":
-        if cohere_chunk.delta.message and cohere_chunk.delta.message.tool_plan is not None:
-            content = cohere_chunk.delta.message.tool_plan
-            state["tool_plan"] += content
-            return content
-    elif cohere_chunk.type == "tool-call-start":
-        if cohere_chunk.delta.message and cohere_chunk.delta.message.tool_calls:
-            tool_call = cohere_chunk.delta.message.tool_calls
+    meta["chunk_type"] = chunk.type
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: Is there a reason you want to save `chunk_type` ? I dont believe its needed.

**Comment by Amnah199** [comment]
@davidsbatista I noticed that `CohereChatGenerator` doesn’t have a `streaming_callback` parameter in its run method.
This prevents streaming when the generator is used within an Agent and a `streaming_callback` is passed to the Agent.
Maybe we can also fix that by adding a new `streaming_callback` parameter in `run` and fixing the below line:
https://github.com/deepset-ai/haystack-core-integrations/blob/0ef151830e2539a0f3f692a2c4de448e4625817c/integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py#L571

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 5]
```diff
@@ -190,144 +177,139 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-                    tool_name=function.name,
-                    arguments={},
-                )
-                # Reset arguments for new tool call
-                state["current_tool_arguments"] = ""
-    elif cohere_chunk.type == "tool-call-delta":
+            if function is not None and function.name is not None:
+                tool_calls = [
+                    ToolCallDelta(
+                        index=0,  # Cohere typically has single tool calls
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: I understand the indexes in this issue are a bit confusing.

Since Cohere supports multiple tool calls, we may need to handle tool_call_index ourselves, similar to how it’s done here:
[Anthropic example](https://github.com/deepset-ai/haystack-core-integrations/blob/aab3afcd4b1e8ce19a1b96215eb340ca114959d0/integrations/anthropic/src/haystack_integrations/components/generators/anthropic/chat/chat_generator.py#L492).

The current implementation skips all tool calls after the first one, which should be fixed.

**Comment by Amnah199** [comment]
I would suggest running an Agent locally with multiple tool calls and checking if the tool calls and tool results are streamed correctly. At the moment there are couple of corrections needed.

Also we can add a test like this one to test correct parsing of tool calls in final chat message. 
https://github.com/deepset-ai/haystack-core-integrations/blob/aab3afcd4b1e8ce19a1b96215eb340ca114959d0/integrations/anthropic/tests/test_chat_generator.py#L333

**Review by Amnah199 — CHANGES_REQUESTED** [review]
Lets fix the tool calls and do another round of review!

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py (hunk 4 above)
Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: ok, I will remove it

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 6]
```diff
@@ -190,144 +177,136 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-        "tool_calls": [],
-        "current_tool_call": None,
-        "current_tool_arguments": "",
-        "captured_meta": {},
-    }
+def _convert_cohere_chunk_to_streaming_chunk(
+    chunk: StreamedChatResponseV2,
+    previous_chunks: List[StreamingChunk],
+    component_info: Optional[ComponentInfo] = None,
+    model: str = "",
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]:
```suggestion
    model: str ,
```

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 7]
```diff
@@ -190,144 +177,136 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
+    """
+    Converts a Cohere streaming response chunk to a StreamingChunk.
 
+    References the Cohere API documentation for the structure of the chunk.
 
-def _process_cohere_chunk(cohere_chunk: StreamedChatResponseV2, state: Dict[str, Any], model: str) -> Optional[str]:
-    """
-    Process a single streamed chat response and update the parsing state.
+    https://docs.cohere.com/reference/chat-stream
+    https://docs.cohere.com/reference/chat#response.body.finish_reason
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py:193]: We can add them as a comment for knowledge sharing internally. Preferably not in docstrings. WDYT

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 8]
```diff
@@ -190,144 +177,136 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
+
+    # early return for invalid chunks
+    if not chunk or not hasattr(chunk, "delta") or chunk.delta is None:
+        return StreamingChunk(
+            content=content,
+            component_info=component_info,
+            index=index,
+            finish_reason=finish_reason,
+            meta=meta,
+        )
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: Hmm what do you mean by invalid chunks? And if its invalid can we not skip streaming this?

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py (hunk 7 above)
Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py:193]: I feel this API can change quickly, so having a reference to the last docs can be beneficial. 

Whoever is using this integration (with Cohere) can quickly look up the API in case something fails, update the code, and submit a PR.

I would rather keep it.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py (hunk 8 above)
Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: something that the code doesn't know how to handle, i.e.: new chunk type due to API changes

Comment (Amnah199) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: I'd stick to the older approach of returning None if the chunk is invalid.
And then checking before appending the streaming chunks.

```
stream_content = _process_cohere_chunk(cohere_chunk=chunk, state=state, model=model)
        if stream_content:
            stream_chunk = StreamingChunk(content=stream_content,
```

Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: makes sense! agree 👍🏽

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 9]
```diff
@@ -190,144 +177,136 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-        if cohere_chunk.delta.message and cohere_chunk.delta.message.tool_calls:
-            tool_call = cohere_chunk.delta.message.tool_calls
+    elif chunk.type == "tool-plan-delta":
+        if chunk.delta.message and chunk.delta.message.tool_plan is not None:
+            content = chunk.delta.message.tool_plan
+            start = len(previous_chunks) == 0
+
+    elif chunk.type == "tool-call-start":
+        if chunk.delta.message and chunk.delta.message.tool_calls:
+            tool_call = chunk.delta.message.tool_calls
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: @davidsbatista I tried running the Agent with multiple tool calls and I couldn't get the second result. Also the streaming is not formatted with `print_streaming_chunk`.
Did you try it with our test code?

Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: yes, it works on my side

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 10]
```diff
@@ -190,144 +177,127 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-            and usage_data.billed_units
-            and usage_data.billed_units.input_tokens is not None
-            and usage_data.billed_units.output_tokens is not None
-        ):
-            prompt_tokens = usage_data.billed_units.input_tokens
-            completion_tokens = usage_data.billed_units.output_tokens
-
-        state["captured_meta"].update(
-            {"usage": {"prompt_tokens": prompt_tokens, "completion_tokens": completion_tokens}}
-        )
```
Comment (anakin87) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: I think it would be better to support both usage formats, because the Cohere API have been frequently switching between them. See https://github.com/deepset-ai/haystack-core-integrations/pull/2149

Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: will do, I was not aware of this , thanks for the feedback 👍🏽

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 11]
```diff
@@ -359,15 +352,17 @@ async def _parse_async_streaming_response(
     """
     Parses Cohere's async streaming chat response into a Haystack ChatMessage.
     """
-    state = _initialize_streaming_state()
+    chunks: List[StreamingChunk] = []
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py:344]: Lets add `tool_call_index` in async function as well.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 12]
```diff
@@ -562,9 +558,16 @@ def run(
...
+        :param streaming_callback: A callback function that is called when a new token is received from the stream.
+            The callback function accepts StreamingChunk as an argument.
+
         :returns: A dictionary with the following keys:
             - `replies`: a list of `ChatMessage` instances representing the generated responses.
         """
+
+        if streaming_callback:
+            self.streaming_callback = streaming_callback
+
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]:
```suggestion

       streaming_callback = select_streaming_callback(
            init_callback=self.streaming_callback,
            runtime_callback=streaming_callback,
            requires_async=False,
        )

```

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 13]
```diff
@@ -625,9 +629,14 @@ async def run_async(
...
         :param tools: A list of tools for which the model can prepare calls. If set, it will override
             the `tools` parameter set during component initialization.
+        :param streaming_callback: A callback function that is called when a new token is received from the stream.
         :returns: A dictionary with the following keys:
             - `replies`: a list of `ChatMessage` instances representing the generated responses.
         """
+
+        if streaming_callback:
+            self.streaming_callback = streaming_callback
+
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]:
```suggestion
        streaming_callback = select_streaming_callback(
            init_callback=self.streaming_callback,
            runtime_callback=streaming_callback,
            requires_async=True,
        )

```

Comment (Amnah199) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: There is a similar statement on line 656. Delete that as well.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 14]
```diff
@@ -190,104 +178,103 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
+    # Initialize default values
+    content = ""
+    index = getattr(chunk, "index", None)
+    start = False
+    finish_reason = None
+    tool_calls = None
+    meta = {"model": model}
+
+    if not chunk or not hasattr(chunk, "delta") or chunk.delta is None:
+        pass
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: Also we need to remove this if we want the tool-call-end to be considered.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 15]
```diff
@@ -190,104 +178,103 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
-        "response_text": "",
-        "tool_plan": "",
-        "tool_calls": [],
-        "current_tool_call": None,
-        "current_tool_arguments": "",
-        "captured_meta": {},
-    }
+def _convert_cohere_chunk_to_streaming_chunk(
+    chunk: StreamedChatResponseV2,
+    previous_chunks: List[StreamingChunk],
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: Lets remove `previous_chunks` param. I tried that locally and this is basically not doing much. I dont think we need it.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py (hunk 12 above)
Comment (davidsbatista) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: The line above that you propose to remove is the `streaming_callback` coming from the run() method. It is to overwrite the one given at init time. If we remove it, passing a `streaming_callback` at run time will have no effect.

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 16]
```diff
@@ -359,21 +344,29 @@ async def _parse_async_streaming_response(
...
-        stream_content = _process_cohere_chunk(cohere_chunk=chunk, state=state, model=model)
-        if stream_content:
-            stream_chunk = StreamingChunk(content=stream_content, component_info=component_info)
-            await streaming_callback(stream_chunk)
+        if chunk.type == "tool-call-start":
+            tool_call_index += 1
+
+        streaming_chunk = _convert_cohere_chunk_to_streaming_chunk(
+            chunk=chunk, component_info=component_info, model=model
+        )
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: We need to pass tool_call_index here.

```suggestion
        streaming_chunk = _convert_cohere_chunk_to_streaming_chunk(
            chunk=chunk, component_info=component_info, model=model,
            tool_call_index=tool_call_index
        )
```

File: integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py [hunk 17]
```diff
@@ -190,104 +178,96 @@ def _parse_response(chat_response: ChatResponse, model: str) -> ChatMessage:
...
+            if function is not None and function.name is not None:
+                tool_calls = [
+                    ToolCallDelta(
+                        index=tool_call_index,
+                        id=tool_call.id,
+                        tool_name=function.name,
+                        arguments=None,
+                    )
+                ]
+                index = +1  # to adjust for the print_streaming_chunk condition
```
Comment (Amnah199) [inline integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: @sjrl maybe we can use some suggestion here. 
Here is how cohere generates indexes:

<p style="border: 0px solid; box-sizing: inherit; margin: 0px 0px 1.25em; padding: 0px; outline-color: rgba(0, 0, 0, 0); color: rgb(255, 255, 255); font-family: CohereText, ui-sans-serif, system-ui, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: color(display-p3 0 0 0 / 0.05); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">If a content delta is passed first and then a tool call delta is passed, both will have an index of zero because the index is used to identify the sequence of events within the same type of event. Here's a breakdown of how this works:</p><ol style="border: 0px solid; box-sizing: inherit; margin: 1.25em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); list-style: decimal; padding-inline-start: 1.625em; color: rgb(255, 255, 255); font-family: CohereText, ui-sans-serif, system-ui, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: color(display-p3 0 0 0 / 0.05); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><li style="border: 0px solid; box-sizing: inherit; margin: 0.5em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); padding-inline-start: 0.375em;"><p style="border: 0px solid; box-sizing: inherit; margin: 1.25em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0);"><strong style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); font-weight: 600; color: rgb(255, 255, 255);">Content Delta</strong>:</p><ul style="border: 0px solid; box-sizing: inherit; margin: 0.75em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); list-style: disc; padding-inline-start: 1.625em;"><li style="border: 0px solid; box-sizing: inherit; margin: 0.5em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); padding-inline-start: 0.375em;">The<span> </span><code style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); font-family: CohereMono, ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, &quot;Liberation Mono&quot;, &quot;Courier New&quot;, monospace; font-feature-settings: normal; font-size: 0.875em; font-variation-settings: normal; background-color: rgb(0, 0, 0); border-radius: 4px; padding-block: 2px; padding-inline: 4px; --tw-ring-shadow: inset/**/0 0 0 calc(1px + 0px)rgba(41, 41, 41, 1); box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgb(41, 41, 41) 0px 0px 0px 1px inset, rgba(0, 0, 0, 0) 0px 0px 0px 0px; --tw-ring-color: rgba(41, 41, 41, 1); --tw-ring-inset: inset; font-weight: 500 !important; color: color(display-p3 0.8235 0.9961 0.9098 / 0.763);">content-delta</code><span> </span>event is emitted whenever the next chunk of text comes back from the model.</li><li style="border: 0px solid; box-sizing: inherit; margin: 0.5em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); padding-inline-start: 0.375em;">Each<span> </span><code style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); font-family: CohereMono, ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, &quot;Liberation Mono&quot;, &quot;Courier New&quot;, monospace; font-feature-settings: normal; font-size: 0.875em; font-variation-settings: normal; background-color: rgb(0, 0, 0); border-radius: 4px; padding-block: 2px; padding-inline: 4px; --tw-ring-shadow: inset/**/0 0 0 calc(1px + 0px)rgba(41, 41, 41, 1); box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgb(41, 41, 41) 0px 0px 0px 1px inset, rgba(0, 0, 0, 0) 0px 0px 0px 0px; --tw-ring-color: rgba(41, 41, 41, 1); --tw-ring-inset: inset; font-weight: 500 !important; color: color(display-p3 0.8235 0.9961 0.9098 / 0.763);">content-delta</code><span> </span>event is assigned an index starting from 0.</li><li style="border: 0px solid; box-sizing: inherit; margin: 0.5em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); padding-inline-start: 0.375em;">Example:<div class="not-prose bg-card-background border-card-border rounded-3 shadow-card-grayscale group relative mb-6 mt-4 flex w-full border" style="border: 1px solid rgb(77, 77, 77); box-sizing: inherit; margin: 1rem 0px 1.5rem; padding: 0px; outline-color: rgba(0, 0, 0, 0); position: relative; display: flex; width: 518px; border-radius: 12px; background-color: rgb(0, 0, 0); box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(0, 0, 0, 0) 0px 0px 0px 0px, rgba(255, 255, 255, 0.07) 0px 1px 2px 0px; --tw-shadow: 0 1px 2px #ffffff12;"><pre class="code-block-root not-prose" tabindex="0" style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); font-family: CohereMono, ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, &quot;Liberation Mono&quot;, &quot;Courier New&quot;, monospace; font-feature-settings: normal; font-size: 1em; font-variation-settings: normal; counter-reset: code-block-line 0; display: flex; flex-direction: column; overflow: hidden; width: 516px; background-color: unset !important;"><div dir="ltr" class="fern-scroll-area" style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); overflow: hidden; z-index: 0; display: flex; flex-direction: column; height: 35.5px; width: 516px; flex-shrink: 1; min-height: 0px; position: relative; --radix-scroll-area-corner-width: 0px; --radix-scroll-area-corner-height: 0px;"><div data-radix-scroll-area-viewport="" class="fern-scroll-area-viewport" data-scrollbars="both" style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); display: flex; flex-direction: column; height: 35.5px; width: 516px; border-radius: inherit; flex-shrink: 1; min-height: 0px; scrollbar-width: none; overflow: scroll;"><div style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); display: table; flex-grow: 1; min-width: 100%;"><code class="code-block text-xs" style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); font-family: CohereMono, ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, &quot;Liberation Mono&quot;, &quot;Courier New&quot;, monospace; font-feature-settings: normal; font-size: 12px; font-variation-settings: normal; --tw-leading: 1.625; line-height: 1.625;"><div class="code-block-inner" style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); padding-block: 8px;">
type='content-delta' index=0 delta=ChatContentDeltaEventDelta(message=ChatContentDeltaEventDeltaMessage(content=ChatContentDeltaEventDeltaMessageContent(text='Hello')))
--


</div></code></div></div></div></pre><button class="focus-visible:ring-(color:--accent) rounded-2 inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium hover:transition-none focus-visible:outline-none focus-visible:ring-1 disabled:pointer-events-none disabled:opacity-50 [&amp;_svg]:pointer-events-none [&amp;_svg]:size-4 [&amp;_svg]:shrink-0 text-(color:--grayscale-a11) hover:bg-(color:--accent-a3) hover:text-(color:--accent-11) pointer-coarse:size-9 size-7 fern-copy-button group fern-copy-button absolute z-20 opacity-0 backdrop-blur transition group-hover:opacity-100 right-2 top-2" data-state="closed" style="border: 0px solid; box-sizing: inherit; margin: 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); background-color: rgba(0, 0, 0, 0); border-radius: 8px; color: rgba(255, 255, 255, 0.686); font-style: inherit; font-variant: inherit; font-weight: 500; font-stretch: inherit; font-size: 0.875rem; line-height: 1.25rem; font-family: inherit; font-optical-sizing: inherit; font-size-adjust: inherit; font-kerning: inherit; font-feature-settings: inherit; font-variation-settings: inherit; letter-spacing: inherit; opacity: 0; appearance: button; --tw-backdrop-blur: blur(8px); backdrop-filter: blur(8px); height: 28px; width: 28px; top: 8px; position: absolute; right: 8px; z-index: 20; display: inline-flex; align-items: center; justify-content: center; gap: 0.5rem; --tw-font-weight: 500; white-space: nowrap; transition-property: color, background-color, border-color, text-decoration-color, fill, stroke, opacity, box-shadow, transform, filter, backdrop-filter, -webkit-backdrop-filter; transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1); transition-duration: 150ms;"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="lucide lucide-copy"><rect width="14" height="14" x="8" y="8" rx="2" ry="2"></rect><path d="M4 16c-1.1 0-2-.9-2-2V4c0-1.1.9-2 2-2h10c1.1 0 2 .9 2 2"></path></svg></button></div></li></ul></li></ol><p style="border: 0px solid; box-sizing: inherit; margin: 1.25em 0px; padding: 0px; outline-color: rgba(0, 0, 0, 0); color: rgb(255, 255, 255); font-family: CohereText, ui-sans-serif, system-ui, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: left; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: color(display-p3 0 0 0 / 0.05); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Same pattern is followed by Tool Calls. Since the indexes are specific to the type of event (content or tool call), they reset independently.</p> 


For example:
```
First content delta: index=0
First tool call delta: index=0 (independent of content delta)
Second content delta: index=1
```

Comment (Amnah199) [reply integrations/cohere/src/haystack_integrations/components/generators/cohere/chat/chat_generator.py]: For `print_streaming_chunk` to format properly we need index > 0 ([here](https://github.com/deepset-ai/haystack/blob/db661df31319ef2e96928efd9ed747d8c94e56d2/haystack/components/generators/utils.py#L28)) and or the index should be less than tool_call_index ([here](https://github.com/deepset-ai/haystack/blob/db661df31319ef2e96928efd9ed747d8c94e56d2/haystack/components/generators/utils.py#L43)).

The current fix with` index=+1` works but I dont think its the best way to do it as it makes `StreamingChunk.index` inconsistent.

File: integrations/cohere/tests/test_chat_generator.py [hunk 18]
```diff
@@ -631,3 +521,517 @@ def test_pipeline_with_cohere_chat_generator(self):
...
+                },
+            },
+        ),
+    ]
+
+
+class TestCohereChunkConversion:
+    def test_convert_cohere_chunk_to_streaming_chunk_complete_sequence(self, cohere_chunks, expected_streaming_chunks):
+        # TODO: the indexes are not correctly checked because of missing global_index
+        # all the streaming chunks have index 0, but the expected indexes increase with tool calls.
```
Comment (Amnah199) [inline integrations/cohere/tests/test_chat_generator.py:884]: @davidsbatista I left a comment here because with the current way of testing the resulting StreamingChunk always have `index = 0`. For a thorough testing, we can take a different approach later.
